File: python/fmvpu/amlet/alu_lite_instruction.py
from dataclasses import dataclass
from typing import List, Tuple
from enum import IntEnum

from fmvpu.control_structures import pack_fields_to_words, unpack_words_to_fields, calculate_total_width, pack_fields_to_int


class ALULiteModes(IntEnum):
    NONE = 0
    ADD = 1
    ADDI = 2
    SUB = 3
    SUBI = 4
    MULT = 5
    MULT_ACC = 6
    MULT_ACC_INIT = 7
    EQ = 8
    GTE = 9
    LTE = 10
    NOT = 11
    AND = 12
    OR = 13
    RESERVED14 = 14
    RESERVED15 = 15
    SHIFT_L = 16
    SHIFT_R = 17
    RESERVED18 = 18
    RESERVED19 = 19
    RESERVED20 = 20
    RESERVED21 = 21
    RESERVED22 = 22
    RESERVED23 = 23
    RESERVED24 = 24
    RESERVED25 = 25
    RESERVED26 = 26
    RESERVED27 = 27
    RESERVED28 = 28
    RESERVED29 = 29
    RESERVED30 = 30
    JUMP = 31


@dataclass
class ALULiteInstruction:
    """ALU Lite instruction for amlet (address calculations)"""
    mode: ALULiteModes = ALULiteModes.NONE
    src1: int = 0  # Source 1 register (a-type register)
    src2: int = 0  # Source 2 register (a-type register)
    dst: int = 0   # Destination register (b-type register)
    
    @classmethod
    def get_width(cls, params) -> int:
        """Get the bit width of this instruction type based on parameters"""
        return calculate_total_width(cls.get_field_specs(params))
    
    @classmethod
    def get_field_specs(cls, params) -> List[Tuple[str, int]]:
        """Get field specifications for bit packing."""
        return [
            ('mode', 5),  # 5 bits to support up to 31
            ('src1', params.a_reg_width),
            ('src2', params.a_reg_width),
            ('dst', params.b_reg_width),
        ]
    
    def encode(self, params) -> int:
        """Encode to instruction bits"""
        field_specs = self.get_field_specs(params)
        return pack_fields_to_int(self, field_specs)
    
    @classmethod
    def from_word(cls, word: int) -> 'ALULiteInstruction':
        """Parse instruction from word"""
        field_specs = cls.get_field_specs()
        field_values = unpack_words_to_fields([word], field_specs, word_width=32)
        return cls(**field_values)

File: src/main/scala/fmvpu/amlet/ALULiteInstruction.scala
package fmvpu.amlet

import chisel3._
import chisel3.util._


object ALULiteInstr {

  object Modes extends ChiselEnum {
    val None = Value(0.U)
    val Add = Value(1.U)       
    val Addi = Value(2.U)
    val Sub = Value(3.U)
    val Subi = Value(4.U)
    val Mult = Value(5.U)
    val MultAcc = Value(6.U)
    val MultAccInit = Value(7.U)
  
    val Eq = Value(8.U)
    val Gte = Value(9.U)
    val Lte = Value(10.U)
    val Not = Value(11.U)
    val And = Value(12.U)
    val Or = Value(13.U)
    val Reserved14 = Value(14.U)
    val Reserved15 = Value(15.U)
  
    val ShiftL = Value(16.U)
    val ShiftR = Value(17.U)
    val Reserved18 = Value(18.U)
    val Reserved19 = Value(19.U)
    val Reserved20 = Value(20.U)
    val Reserved21 = Value(21.U)
    val Reserved22 = Value(22.U)
    val Reserved23 = Value(23.U)
    val Reserved24 = Value(24.U)
    val Reserved25 = Value(25.U)
    val Reserved26 = Value(26.U)
    val Reserved27 = Value(27.U)
    val Reserved28 = Value(28.U)
    val Reserved29 = Value(29.U)
    val Reserved30 = Value(30.U)
    val Jump = Value(31.U)
  }
  
  class Base(params: AmletParams) extends Instr.Base(params) {
    val mode = Modes()
    val src1 = params.aReg()
    val src2 = params.aReg()
    val dst = params.bReg()
  }
  
  class Resolving(params: AmletParams) extends Instr.Resolving(params) {
    val mode = Modes()
    val src1 = new ATaggedSource(params)
    val src2 = new ATaggedSource(params) 
    val mask = new MaskInfo(params)
    val dst = new BTaggedReg(params)

    def isResolved(): Bool = {
      src1.resolved && 
      src2.resolved && 
      mask.resolved
    }

    def isMasked(): Bool = {
      mask.resolved && mask.getData
    }

    def resolve(): Resolved = {
      val resolved = Wire(new Resolved(params))
      resolved.mode := mode
      resolved.src1 := src1.getData
      resolved.src2 := src2.getData
      resolved.dst := dst
      resolved
    }

    def update(writes: ResultBus): Resolving = {
      val resolving = Wire(new Resolving(params))
      resolving.mode := mode
      resolving.src1 := src1.update(writes)
      resolving.src2 := src2.update(writes)
      resolving.mask := mask.update(writes)
      resolving.dst := dst
      resolving
    }
  }

  class Resolved(params: AmletParams) extends Instr.Resolved(params) {
    val mode = Modes()
    val src1 = params.aWord()
    val src2 = params.aWord()
    val dst = new BTaggedReg(params)
  }

}
